test/client.js

```diff
@@ -252,5 +252,70 @@ describe('MqttClient', function () {
         }
       }, 2000)
     })
+
+    it('shoud not send the same packet multiple times on a flaky connection', function (done) {
+      this.timeout(3500)
+
+      var killCount = 4
+      var subIds = {}
+      var client = mqtt.connect({
+        port: port + 46,
+        host: 'localhost',
+        connectTimeout: 350,
+        reconnectPeriod: 300
+      })
+
+      var server2 = new Server(function (client) {
+        client.on('connect', function (packet) {
+          if (packet.clientId === 'invalid') {
+            client.connack({returnCode: 2})
+          } else {
+            client.connack({returnCode: 0})
+          }
+        })
+      }).listen(port + 46)
+
+      server2.on('client', function (c) {
+        client.subscribe('topic')
+        c.on('subscribe', function (packet) {
+          if (killCount) {
+            // Kill the first few sub attempts to simulate a flaky connection
+            killCount--
+            c.destroy()
+          } else {
+            // Keep track of acks
+            if (!subIds[packet.messageId]) {
+              subIds[packet.messageId] = 0
+            }
+            subIds[packet.messageId]++
+
+            c.suback({
+              messageId: packet.messageId,
+              granted: packet.subscriptions.map(function (e) {
+                return e.qos
+              })
+            })
+          }
+        })
+      })
+
+      setTimeout(function () {
+        var duplicateId
+        Object.keys(subIds).some(function (messageId) {
+          if (subIds[messageId] > 1) {
+            duplicateId = subIds[messageId]
+            return true
+          } else {
+            return false
+          }
+        })
+        if (duplicateId) {
+          done(new Error('Multiple duplicate acked subscriptions received for messageId ' + duplicateId))
+        } else {
+          done()
+        }
+        client.end(true)
+      }, 3000)
```

Comment:
I am really scared about timeouts in tests, as they are often a source of flakyness.
Can we remove it?